### PR TITLE
[MXNET -1004] Poisson NegativeLog Likelihood loss

### PR DESCRIPTION
**Pull request by gaurav-gireesh**
## Description ##
The function computes the poisson's negative log likelihood loss that serves as measurement of the loss accrued from a Poisson regression prediction made by the model.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)

- [x] Code is well-documented: 

- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] PoissonNLLLoss computation function tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.

 - Below is the link for pytorch's current implementation:
    [PyTorch Implementation] . 
    (https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py)

### CODE REVIEW

**Comment by vandanavk** [comment]
@mxnet-label-bot [pr-work-in-progress]

**Comment by anirudhacharya** [comment]
@gaurav-gireesh what is in progress for this PR? The PR description says that you have completed all the code changes and all the checkboxes are ticked.

File: tests/python/unittest/test_loss.py [hunk 1]
```diff
@@ -185,27 +185,28 @@ def test_l1_loss():
 @with_seed(1234)
 def test_ctc_loss():
     loss = gluon.loss.CTCLoss()
-    l = loss(mx.nd.ones((2,20,4)), mx.nd.array([[1,0,-1,-1],[2,1,1,-1]]))
+    l = loss(mx.nd.ones((2, 20, 4)), mx.nd.array([[1, 0, -1, -1], [2, 1, 1, -1]]))
```
Comment (roywei) [inline tests/python/unittest/test_loss.py]: As discussed, please address pylint issue in another PR, it should be going to master first before this PR

File: tests/python/unittest/test_loss.py [hunk 2]
```diff
@@ -342,12 +343,45 @@ def test_triplet_loss():
...
     loss = mx.sym.make_loss(loss)
-    mod = mx.mod.Module(loss, data_names=('data',), label_names=('pos','neg'))
+    mod = mx.mod.Module(loss, data_names=('data',), label_names=('pos', 'neg'))
     mod.fit(data_iter, num_epoch=200, optimizer_params={'learning_rate': 0.01},
             initializer=mx.init.Xavier(magnitude=2), eval_metric=mx.metric.Loss(),
             optimizer='adam')
     assert mod.score(data_iter, eval_metric=mx.metric.Loss())[0][1] < 0.05
 
+@with_seed()
+def test_poisson_nllloss():
```
Comment (roywei) [inline tests/python/unittest/test_loss.py]: Please separate into 3 unit tests clearly stating what condition are you testing.

File: python/mxnet/gluon/loss.py [hunk 3]
```diff
@@ -706,3 +707,64 @@ def hybrid_forward(self, F, pred, positive, negative):
...
+    def hybrid_forward(self, F, pred, target, sample_weight=None, epsilon=1e-08):
+        target = _reshape_like(F, target, pred)
+        if self._from_logits:
+            loss = F.exp(pred) - target * pred
+        else:
+            loss = pred - target * F.log(pred + epsilon)
+        if self._compute_full:
+            stirling_factor = target * F.log(target)- target + 0.5 * F.log(2 * target * np.pi)
+            target_gt_1 = target > 1
+            stirling_factor *= target_gt_1
```
Comment (roywei) [inline python/mxnet/gluon/loss.py]: Add a comment explaining why we need this

**Comment by roywei** [comment]
> @gaurav-gireesh what is "in progress" for this PR? The PR description says that you have completed all the code changes and all the checkboxes are ticked.

@anirudhacharya  there was an issue with CI and pylin, we are looking into it.

**Comment by gaurav-gireesh** [comment]
> @gaurav-gireesh what is "in progress" for this PR? The PR description says that you have completed all the code changes and all the checkboxes are ticked.

@anirudhacharya It was a miss Anirudha. Corrected that and added a few test cases for Stirling approximation too.

File: tests/python/unittest/test_loss.py (hunk 2 above)
Comment (gaurav-gireesh) [reply tests/python/unittest/test_loss.py]: @Lai Yes, I have separated the use cases commenting on what exactly is being tested. I have not split into different functions so as to keep consistency with other loss functions tests.
Thank you for your comment.

File: python/mxnet/gluon/loss.py (hunk 3 above)
Comment (gaurav-gireesh) [reply python/mxnet/gluon/loss.py]: Thanks!
Yes, I have mentioned the use of the parameter 'compute_full' in function docstring, which explains why the calculation is needed.

File: python/mxnet/gluon/loss.py [hunk 4]
```diff
@@ -706,3 +707,65 @@ def hybrid_forward(self, F, pred, positive, negative):
...
+        L = \text{pred} - \text{target} * \log(\text{pred}) +\log(\text{target!})
+
+    `pred`, `target` can have arbitrary shape as long as they have the same number of elements.
+
+    Parameters
+    ----------
+    from_logits : boolean, default True
+        indicating whether log(predicted) value has already been computed. If True, the loss is computed as
+        :math:`\exp(\text{pred}) - \text{target} * \text{pred}`, and if False, then loss is computed as
+        :math:`\text{pred} - \text{target} * \log(\text{pred}+\text{epsislon})`.The default value
```
Comment (anirudhacharya) [inline python/mxnet/gluon/loss.py]: these definitions seem to be at odds with the definition given above in line 718

File: tests/python/unittest/test_loss.py [hunk 5]
```diff
@@ -348,6 +348,44 @@ def test_triplet_loss():
...
+    target[:] += mx.nd.abs(min_target)
+
+    Loss = gluon.loss.PoissonNLLLoss(from_logits=True)
+    Loss_no_logits = gluon.loss.PoissonNLLLoss(from_logits=False)
+    #Calculating by brute formula for default value of from_logits = True
+
+    # 1) Testing for flag logits = True
+    brute_loss = np.mean(np.exp(pred.asnumpy()) - target.asnumpy() * pred.asnumpy())
+    loss_withlogits = Loss(pred, target)
+    assert_almost_equal(brute_loss, loss_withlogits.asscalar())
```
Comment (anirudhacharya) [inline tests/python/unittest/test_loss.py]: why not test this loss function using the Module API with `module.fit` and `module.score`

File: python/mxnet/gluon/loss.py [hunk 6]
```diff
@@ -706,3 +707,65 @@ def hybrid_forward(self, F, pred, positive, negative):
...
         return _apply_weighting(F, loss, self._weight, None)
+
+
+class PoissonNLLLoss(Loss):
+    r"""For a target (Random Variable) in a Poisson distribution, the function calculates the Negative
+    Log likelihood loss.
+    PoissonNLLLoss measures the loss accrued from a poisson regression prediction made by the model.
+
+    .. math::
+        L = \text{pred} - \text{target} * \log(\text{pred}) +\log(\text{target!})
```
[the remark this answers is not in the document]
Comment (gaurav-gireesh) [reply python/mxnet/gluon/loss.py]: @anirudhacharya The "!" is for factorial of the target value. The formula to calculate the probability of the target value with the mean of the poisson distribution contains a factorial term in the denominator which is approximated by computing the Stirling approximation. Taking a log of the formula reduces to the form mentioned in the documentation.
Some wikipedia pages that I have found helpful are:
[Poisson regression](https://en.wikipedia.org/wiki/Poisson_regression) and
[Poisson distribution](https://en.wikipedia.org/wiki/Poisson_distribution).
Also, I have attached a link for reference to PyTorch's implementation of the loss function in the description of the PR.

File: tests/python/unittest/test_loss.py (hunk 5 above)
Comment (gaurav-gireesh) [reply tests/python/unittest/test_loss.py]: Thank you for the suggestion. It is a valid point however, getting a synthetic data set to behave in a way that Random input X's are correlated to a target ~ PoissonDistribution(target belonging to a Poisson Distribution) is difficult to gather. However, I have tried with random data sets and trained the model but the loss gradient was not observed. This could be incrementally contributed if I come across such a dataset which shows the loss gradient with epochs as other models like LogisticRegression, Linear Regression and so on.
The formula to compute a loss value however, is implemented in the function which can be tested by raw calculations. This is also something that we see in unit tests of loss functions such as : test_bce_loss(Binary Cross Entropy).

File: python/mxnet/gluon/loss.py (hunk 4 above)
Comment (gaurav-gireesh) [reply python/mxnet/gluon/loss.py]: Yes the formulae are different. Following are some points:
1)  The factorial term gets dropped in most calculations. However, there is a way to approximate it using Stirling factor/approximation. We can use **_compute_full_** flag to be set to True to incorporate this approximation in our loss value computation,
2) from_logits flag set to true or false can render the formula to change depending upon the prediction is already logged ( log pred) or not.

File: python/mxnet/gluon/loss.py [hunk 7]
```diff
@@ -23,8 +23,8 @@
            'SigmoidBinaryCrossEntropyLoss', 'SigmoidBCELoss',
            'SoftmaxCrossEntropyLoss', 'SoftmaxCELoss',
            'KLDivLoss', 'CTCLoss', 'HuberLoss', 'HingeLoss',
-           'SquaredHingeLoss', 'LogisticLoss', 'TripletLoss']
-
+           'SquaredHingeLoss', 'LogisticLoss', 'TripletLoss', 'PoissonNLLLoss']
```
Comment (sandeep-krishnamurthy) [inline python/mxnet/gluon/loss.py]: nit: please add line space

File: tests/python/unittest/test_loss.py [hunk 8]
```diff
@@ -348,6 +348,44 @@ def test_triplet_loss():
...
+    #2) Testing for flag logits = False
+    loss_no_logits = Loss_no_logits(pred, target)
+    np_loss_no_logits = np.mean(pred.asnumpy() - target.asnumpy() * np.log(pred.asnumpy() + 1e-08))
+    if np.isnan(loss_no_logits.asscalar()):
+        assert_almost_equal(np.isnan(np_loss_no_logits), np.isnan(loss_no_logits.asscalar()))
+    else:
+        assert_almost_equal(np_loss_no_logits, loss_no_logits.asscalar())
+
+    #3) Testing for Sterling approximation
+    np_pred = np.random.uniform(1, 5, (2, 3))
```
Comment (sandeep-krishnamurthy) [inline tests/python/unittest/test_loss.py]: Please add tests for hybridized version as well.

File: tests/python/unittest/test_loss.py (hunk 5 above)
Comment (gaurav-gireesh) [reply tests/python/unittest/test_loss.py]: Unit test added.

File: tests/python/unittest/test_loss.py (hunk 8 above)
Comment (gaurav-gireesh) [reply tests/python/unittest/test_loss.py]: > Please add tests for hybridized version as well.

Added the unit test for it.

File: python/mxnet/gluon/loss.py (hunk 7 above)
Comment (gaurav-gireesh) [reply python/mxnet/gluon/loss.py]: Added it. Thanks for pointing out.

**Comment by gaurav-gireesh** [comment]
@mxnet-label-bot [pr-awaiting-review]

**Comment by gaurav-gireesh** [comment]
@sandeep-krishnamurthy This PR is awaiting review, so I request you to change the label.
Thanks!
